### DISCUSSION TITLE
Independent Component Analysis (ICA)

### DISCUSSION THREAD

**Discussion by pfeatherstone** [davisking/dlib#2867] [Q&A]
Is there a way to do ICA in dlib using some of the traditional machine learning tools? Something that mirrors https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.FastICA.html

**Comment by pfeatherstone** [comment]
Specifically, I'm trying to write an algorithm with dlib which extract pulse rate from a video. The algorithms typically do:
* detect face
* average RGB pixels from face excluding eyes, mouth, hair
* track (which gives you R(t), G(t), B(t)
* Perform ICA on (R(t), G(t), B(t))  which gives you (X1(t), X2(t), X3(t)) which are 3 linearly independent channels
* Perform FFT on **X**(t) and take max peak
* Peak gives you pulse rate

Something like that. The bit I'm missing is ICA.

**Reply by arrufat** [reply]
Would PCA work for you? Or is it too restrictive? Instead of just linearly independent channels, it would find 3 orthogonal channels.

**Reply by pfeatherstone** [reply]
My understanding of PCA is that it compresses information, whereas I want to decompose. I'm most likely wrong. Maybe PCA will do what i want. Either way, i think i want it to use/train/optimize on the whole time series, not on pixel by pixel basis.

**Reply by arrufat** [reply]
It only compresses if you don't use all dimensions. If you use all dimensions, there's no compression.

**Reply by arrufat** [reply]
Basically, it will put an axis on the dimension that has the largest variance (or energy) and start appending dimensions orthogonally to the previous dimensions from highest to lowest variance.

**Reply by davisking** [reply]
Na you need to code up something for ICA. The linear algebra tooling should be useful. I don't think the solvers or kernel stuff would help you though.

**Reply by pfeatherstone** [reply]
![image](https://github.com/davisking/dlib/assets/45853521/aa2d90d6-a02e-4955-a777-7f147a5d12d2)

I don't know enough about traditional machine learning stuff. This gives a good comparison of PCA vs ICA.

**Reply by arrufat** [reply]
Yes, it seems like PCA is not doing what you need. It's not suitable for this problem, since the sources are not orthogonal, it can't separate them.

**Comment by pfeatherstone** [comment]
It feels like dlib already has all the necessary tools to do this. E.g. using the kernel solvers

**Reply by pfeatherstone** [reply]
Maybe `linearly_independent_subset_finder` is relevant here. Not sure. I'll read the examples but this isn't my area of expertise.

**Comment by pfeatherstone** [comment]
https://en.wikipedia.org/wiki/FastICA spells the algorithm out. Could use this over trying to copy-paste sklearn or some other implementation. Hopefully, the unit tests would show near identical results when using dlib or sklearn on same input data.

**Reply by pfeatherstone** [reply]
We could also eventually add ILRMA, FastMNMF and other blind source separation algorithms which are related to ICA. I've used FastMNMF in the past for solving cocktail party problems but never thought more generally of other use cases.
